### PR TITLE
[WIP] ESPNet2 Recipe Update -- commonvoice, babel, ami

### PR DESCRIPTION
**Pull request by ftshijt**
the PR is related to #1490.

- [x] ami
- [x] babel
- [x] commonvoice

A preliminary version

### CODE REVIEW

**Comment by ftshijt** [comment]
It seems that the current espnet2 version does not support segment-level data, doesn't it? I'm experiencing an error from ESPnetDataset for Babel datasets, because it defaults the wav.scp and text to build the loader_dict instead of considering the segment info. @kamo-naoyuki Am I get it the right way?

**Comment by kamo-naoyuki** [comment]
`format_wav_scp.sh` supports it.

https://github.com/espnet/espnet/blob/48ce19107633297e149d293ab7b4aa46bc95038a/egs2/TEMPLATE/asr1/scripts/audio/format_wav_scp.sh#L91-L107

Sorry, It should be used here, but I forgot it.
https://github.com/espnet/espnet/blob/48ce19107633297e149d293ab7b4aa46bc95038a/egs2/TEMPLATE/asr1/asr.sh#L294-L300

Please change like this and test with it.

```bash
_opts=
if [ -e data/"${dset}"/segments ]; then
    _opts+="--segments data/${dset}/segments "
fi
scripts/audio/format_wav_scp.sh --nj "${nj}" --cmd "${train_cmd}" \ 
         --audio-format "${audio_format}" --fs "${fs}" \ 
         ${_opts}
         "data/${dset}/wav.scp" "${data_feats}/${dset}" 

```

**Comment by sw005320** [comment]
I'll merge #1513 soon.

File: egs2/TEMPLATE/asr1/pyscripts/audio/format_wav_scp.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ def utt2ref_channels(x, d=utt2ref_channels_dict) -> Tuple[int, ...]:
     if args.segments is not None:
         # Note: kaldiio supports only wav-pcm-int16le file.
         loader = kaldiio.load_scp(args.scp, segments=args.segments)
-        with SoundScpWriter(args.outdir, args.name,
+        with SoundScpWriter(args.outdir, Path(args.outdir) / f'{args.name}.scp',
```
Comment (ftshijt) [inline egs2/TEMPLATE/asr1/pyscripts/audio/format_wav_scp.py]: There is still a minor bug for the segment version to take effect.  I fixed here, please have a check @kamo-naoyuki Thank you!

Comment (kamo-naoyuki) [reply egs2/TEMPLATE/asr1/pyscripts/audio/format_wav_scp.py]: Thank you! Looks good.
I forgot that I changed the interface of SoundScpWriter.

**Comment by ftshijt** [comment]
Hi, @kamo-naoyuki .  I'm thinking about adding the `remove_longshortdata.sh ` in the asr.sh. I cannot make it in `data.sh` cause the frame information is necessary to process.

To make it consistent with others' new recipe, I may introduce another parameter for asr.sh, which controls if the remove script is taking effect. Is that OK?

**Comment by kamo-naoyuki** [comment]
> To make it consistent with others' new recipe, I may introduce another parameter for asr.sh, which controls if the remove script is taking effect. Is that OK?

I see, it is necessary.  Thank you for telling me. I have some thoughts.

Do you have an idea to handle raw wave and fbank both? We need to implement new `remove_longshortdata.sh` supporting  both.

And, also I think, there several possibilities. One is, as you said, creating such stage to remove them, and the other is on-the-fly filtering at training process.

For latter pattern, we can implement such mechanism here: 
https://github.com/espnet/espnet/blob/ea6d2e2d9fb3ef020cd871c7443b637d6754c786/espnet2/train/batch_sampler.py#L89-L105
and
https://github.com/espnet/espnet/blob/ea6d2e2d9fb3ef020cd871c7443b637d6754c786/espnet2/train/batch_sampler.py#L209-L218

I'm not sure which is better. The former is more clear and the latter is flexible about the threshold.
What do you think? 
@sw005320 I'd like to ask you, too. 

If you'll select creating new stage to remove utterances, you need to also get the frame length some where.
About `raw`, `format_wav_scp.sh` creates `utt2num_samples` together, but for `fbank` such files are not created now, so I think it's better to insert the stage after `stage7`

 https://github.com/espnet/espnet/blob/ea6d2e2d9fb3ef020cd871c7443b637d6754c786/egs2/TEMPLATE/asr1/asr.sh#L522-L525

`collect stats` dumps the shape information of feature and text data and 
 the samples used in training are finally decided only by `shape_file`, which you can see at `exp/asr_stats/train/*_shape`.

**Comment by ftshijt** [comment]
> `collect stats` dumps the shape information of feature and text data and
> the samples used in training are finally decided only by `shape_file`, which you can see at `exp/asr_stats/train/*_shape`.

Actually, the reason I came up with the `remove_longshortdata` is from the AMI corpus in which I found that there are some wrong empty segments(I check the `segments` file, I believe there are some annotation mistakes). As for processing these audios in Stage 7 (the ASR stats collect), it will raise an error in the SoundFile loader. Hence, it may be better to deal with checking the filtering in the stats collection section. What do you think of it? @kamo-naoyuki

**Comment by kamo-naoyuki** [comment]
> Actually, the reason I came up with the `remove_longshortdata` is from the AMI corpus in which I found that there are some wrong empty segments(I check the `segments` file, I believe there are some annotation mistakes). 

I understand what happened. FYI. Kaldiio, which uses python builtlin wave loader, can handles 0-frame audio, but Soundfile can't load if and, it returns None object if 0frames is loaded.

O.K.  I'll try to add this feature somehow. Actually I should also apply it in voxforge. Wait a little please.

**Comment by sw005320** [comment]
Sorry for my late response.
My preference is to perform it in advance as it is more understandable.

**Comment by ftshijt** [comment]
hi, @kamo-naoyuki , I checked the remove_short_data stage and found that there might be an error with "segments" file. The reason is because there will be a mismatch between segments & wav.scp after the wav_format. 

For me, a temporary solution is to rm the segments file in `dump/raw/train`, which would kind of escape from the problem. Do you have any suggestions?

**Comment by kamo-naoyuki** [comment]
could you show me  the error?

**Comment by ftshijt** [comment]
> 2020-02-04T12:11:59 (asr.sh:348:main) Stage 3: Remove short data: dump/raw/org -> dump/raw
> utils/copy_data_dir.sh: copied data from dump/raw/org/ihm_train to dump/raw/ihm_train
> utils/validate_data_dir.sh: Error: in dump/raw/ihm_train, recording-ids extracted from segments and wav.scp
> utils/validate_data_dir.sh: differ, partial diff is:
> --- /tmp/kaldi.xFhe/recordings  2020-02-04 12:12:17.024279768 -0500
> +++ /tmp/kaldi.xFhe/recordings.wav      2020-02-04 12:12:17.228277789 -0500
> @@ -1,1641 +1,325506 @@
> -sp0.9-AMI_EN2001a_H00
> -sp0.9-AMI_EN2001a_H01
> -sp0.9-AMI_EN2001a_H02
> ...
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0229364_0229459
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0229459_0229845
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0230011_0230278
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0230278_0230313
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0230804_0230825
> +sp1.1-AMI_TS3012d_H03_MTD048ME_0232390_0232439
> [Lengths are /tmp/kaldi.xFhe/recordings=1641 versus /tmp/kaldi.xFhe/recordings.wav=325506]

take AMI as an example @kamo-naoyuki

**Comment by kamo-naoyuki** [comment]
Oh, it can be resolved just removing segments file as you do.
It should be done after format_wav_scp.sh.


(We need ci test for segments file. I'll put)

**Comment by ftshijt** [comment]
@kamo-naoyuki Hi, I found a problem for ci test as follows:
image
The problem is because I reference cmd.sh by `. ./cmd.sh`, but the cmd is not in "local" dir where data.sh lies. Do you have some suggestions on how to deal with it

File: egs2/TEMPLATE/asr1/asr.sh [hunk 2]
```diff
@@ -373,6 +373,15 @@ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
     log "Stage 4: Remove short data: ${data_feats}/org -> ${data_feats}"
 
     for dset in "${train_set}" "${dev_set}" ${eval_sets}; do
+
+        # no need to keep segments & reco2file_and_channel
+        if [ -f ${data_feats}/org/${dset}/segments ]; then
+          rm ${data_feats}/org/${dset}/segments
+        fi
```
Comment (kamo-naoyuki) [inline egs2/TEMPLATE/asr1/asr.sh]: Sorry, I already added at #1564 about segments, but  I didn't noticed for reco2file_and_channel.
This operation must be done only when feats_type=raw.

Could you send another PR as far common fix? And this problem is also related to tts.sh.

Comment (ftshijt) [reply egs2/TEMPLATE/asr1/asr.sh]: please take a look at #1573

**Comment by kamo-naoyuki** [comment]
> The problem is because I reference cmd.sh by `. ./cmd.sh`, but the cmd is not in "local" dir where data.sh lies. Do you have some suggestions on how to deal with it

@ftshijt Thanks. it was fixed now. #1572

**Comment by ftshijt** [comment]
Hi, @kamo-naoyuki  the ci issue seems to be unfixed

**Comment by kamo-naoyuki** [comment]
Did you merge v.0.7.0?

**Comment by ftshijt** [comment]
> Did you merge v.0.7.0?

Sorry to bother you, my fault...

**Comment by kamo-naoyuki** [comment]
No problem. Actually, I'm also not sure.

**Comment by ftshijt** [comment]
Hi, @kamo-naoyuki , could you please take a look at my babel recipe? I've tested it several times on (lang 203, Lao). The training curve is as follows:
image

However, I also tried ESPNet version1 and got the following curve.
image

There are two kinds of stuff that may matters. The first might be the distinct language characters of Lao. They use characters such as ອື ຕົນເອງ ຄື ບໍ່ ແວ່ ມາ ຫຼິນ ນຳ ເຂົາ ແດ່. The other one might be the non-linguistic symbols, there are a lot of lysyms (e.g. <noise>, <silence>,  <hes>, etc.) To handle them, I have generated a nlsyms.txt  and add it to the system. I've looked through the code for the two matters, but do not find anything special. Do you have some ideas on it?

**Comment by ftshijt** [comment]
BTW, It seems I still cannot pass the CI check. The error seems to have something wrong with the CI process itself.

File: egs2/TEMPLATE/asr1/asr.sh (hunk 2 above)
Comment (kamo-naoyuki) [reply egs2/TEMPLATE/asr1/asr.sh]: Remove this changing.

**Comment by kamo-naoyuki** [comment]
> However, I also tried ESPNet version1 and got the following curve.

Could you use xavier uniform? I don't know why but the default initialization gives such training curve. 
https://github.com/espnet/espnet/pull/1601

> There are two kinds of stuff that may matters. The first might be the distinct language characters of Lao. They use characters such as ອື ຕົນເອງ ຄື ບໍ່ ແວ່ ມາ ຫຼິນ ນຳ ເຂົາ ແດ່. The other one might be the non-linguistic symbols, there are a lot of lysyms (e.g. , , , etc.) To handle them, I have generated a nlsyms.txt and add it to the system. I've looked through the code for the two matters, but do not find anything special. Do you have some ideas on it?

Sorry, I couldn't understand what problem happens. The problem is that there are a lot of non-linguistic symbols?

**Comment by ftshijt** [comment]
> > However, I also tried ESPNet version1 and got the following curve.
> 
> Could you use xavier uniform? I don't know why but the default initialization gives such training curve.
> #1601
> 
> > There are two kinds of stuff that may matters. The first might be the distinct language characters of Lao. They use characters such as ອື ຕົນເອງ ຄື ບໍ່ ແວ່ ມາ ຫຼິນ ນຳ ເຂົາ ແດ່. The other one might be the non-linguistic symbols, there are a lot of lysyms (e.g. , , , etc.) To handle them, I have generated a nlsyms.txt and add it to the system. I've looked through the code for the two matters, but do not find anything special. Do you have some ideas on it?
> 
> Sorry, I couldn't understand what problem happens. The problem is that there are a lot of non-linguistic symbols?

Thank you for your advice. I've started training yesterday on your conf. 

Yes, there are frequently non-linguistic symbols (some examples are shown as follows). I suspect they may affect the training. In espnet1, the nlsyms are removed during training. But in espnet2, it seems that it is not removed for scoring.

Screen Shot 2020-02-23 at 22 24 47

**Comment by kamo-naoyuki** [comment]
7

2020年2月24日(月) 12:35 ftshijt <notifications@github.com>:

**Comment by kamo-naoyuki** [comment]
> Yes, there are frequently non-linguistic symbols (some examples are shown as follows). I suspect they may affect the training. In espnet1, the nlsyms are removed during training.

If so, it should be removed at local/data.sh,  but I'm not sure whether it's good or not. In my experience, such nlsyms should be included in training.
I also didn't know that espnet1 removes it. Could you show me where  nlsyms is removed in espnet1?


> But in espnet2, it seems that it is not removed for scoring.

Non linguistic symbols should be removed at scoring https://github.com/espnet/espnet/blob/faad51705c4ee99f3ff77288b817921dee06953c/egs2/TEMPLATE/asr1/asr.sh#L846-L854

Could you investigate why it's not removed?

**Comment by ftshijt** [comment]
> > Yes, there are frequently non-linguistic symbols (some examples are shown as follows). I suspect they may affect the training. In espnet1, the nlsyms are removed during training.
> 
> If so, it should be removed at local/data.sh, but I'm not sure whether it's good or not. In my experience, such nlsyms should be included in training.

Sorry, I do not mean for training, it's scoring.


> I also didn't know that espnet1 removes it. Could you show me where nlsyms is removed in espnet1?
> 
> > But in espnet2, it seems that it is not removed for scoring.
> 
> Non linguistic symbols should be removed at scoring
> 
> https://github.com/espnet/espnet/blob/faad51705c4ee99f3ff77288b817921dee06953c/egs2/TEMPLATE/asr1/asr.sh#L846-L854
> 
> Could you investigate why it's not removed?
The cer computing is removed. The WER is not.

**Comment by kamo-naoyuki** [comment]
Sorry, the removing of nlsyms for word is not implemented now, I'll send PR...

**Comment by ftshijt** [comment]
> Sorry, the removing of nlsyms for word is not implemented now, I'll send PR...

Thanks! BTW, after I tried the xavier uniform, the training line is getting much close to what I expected. Thank you for your suggestion

**Comment by kamo-naoyuki** [comment]
Now nlsyms are removed #1614

**Comment by ftshijt** [comment]
> Now nlsyms are removed #1614

Thanks!

**Comment by ftshijt** [comment]
Hi, @kamo-naoyuki , 

after training on the xavier_uniform init setting (and some other change based on your #1601 
I got the following acc curve. Though the acc achieved from the model seems to similar to the espnet1's. But the decoding result is still not as compatible (80% wer compare to 50% in espnet1)
acc-6-layer
It seems wired comparing to the espnet1's smooth one. Have you ever observed this kind of curve?

For your reference, the espnet1's curve is like this:
acc

**Comment by kamo-naoyuki** [comment]
Hmm, looks odd. Is this fbank pitch and warp-CTC?

**Comment by ftshijt** [comment]
> Hmm, looks odd. Is this fbank pitch and warp-CTC?

It's based on raw. 
For loss, I follow the nearly up-to-data model (I just pulled yesterday), so I figure I'm using the PyTorch CTC (as I notice you are changing it to a default one)

**Comment by kamo-naoyuki** [comment]
You are still using pytorch=1.0? 
I don't the difference between 1.4 and 1.0, but the behaviour of ctc was changed?

**Comment by ftshijt** [comment]
> You are still using pytorch=1.0?
> I don't the difference between 1.4 and 1.0, but the behaviour of ctc was changed?

I see... I'm using pytorch 1.4 for the current setting. I will change the version to have it a try. Thanks

**Comment by kamo-naoyuki** [comment]
The investigation related to warp ctc and pytorch version is not matter of this PR,  so could you test only pytorch=1.4 and pytorch CTC with fbank_pitch? 

Even if we still found bad result in that case, I'll merge your PR for this recipe because I'm planning to beta release for epsnet2 soon.
(so if possible, would you split the finished recipes?)

**Comment by ftshijt** [comment]
The babel is not a large set, I've started to test. One day would be enough for the whole process. So you may directly combine it

**Comment by kamo-naoyuki** [comment]
BTW, espnet1's results also seem not converged in 15 epochs, could you try more epochs? We can possibly get better result.

**Comment by ftshijt** [comment]
acc
It seems that the pytorch14 does not change the performance on dev set a lot. The result is almost the same as pytorch11. I will push my latest results to the three sets.

**Comment by ftshijt** [comment]
> BTW, espnet1's results also seem not converged in 15 epochs, could you try more epochs? We can possibly get better result.

Yes, of course, actually even current result has outperformed (around 10 percent better) the reports in espnet1.

File: egs2/ami/asr1/conf/tuning/train_rnn.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# hybrid CTC/attention
+e2e_conf:
+    ctc_weight: 0.2
+    lsm_weight: 0.05
+
+# minibatch related
+batch_size: 30
+max_length:
+    - 800  # if input length  > maxlen_in, batchsize is automatically reduced
+    - 150 # if output length > maxlen_out, batchsize is automatically reduced
```
Comment (kamo-naoyuki) [inline egs2/ami/asr1/conf/tuning/train_rnn.yaml]: Max length cannot be specified in config file because it should be changed depending on feats_type.

File: egs2/ami/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#  Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+stage=0
+stop_stage=10
+nj=32
+
```
Comment (kamo-naoyuki) [inline egs2/ami/asr1/run.sh]: You don't need to specify stage, stop_stage, nj  because they are not task depended.
I'd like to make all recipes to follow asr.sh as possible.

**Comment by kamo-naoyuki** [comment]
Thanks! Did you use egs2/babel/asr1/conf/train.yaml?

The babel recipe in espnet1 uses label-smoothing:

https://github.com/espnet/espnet/blob/24b07c73e0c3b98177d376cb7e0e5c936ed78a53/egs/babel/asr1/conf/train.yaml#L20-L26

The default of lsm_weight=0 in espnet2, so could you also try lsm_weight=0.05 and ctc_weight=0.5? Note that ESPnet2 doesn't implement unigram lsm.

Could you advice me @sw005320 ? 
I'm not sure whether this is a sign of some bugs. I don't know the effect of lsm and what kind of data babel is.

**Comment by ftshijt** [comment]
> Thanks! Did you use egs2/babel/asr1/conf/train.yaml?
> 
> The babel recipe in espnet1 uses label-smoothing:
> 
> https://github.com/espnet/espnet/blob/24b07c73e0c3b98177d376cb7e0e5c936ed78a53/egs/babel/asr1/conf/train.yaml#L20-L26
> 
> The default of lsm_weight=0 in espnet2, so could you also try lsm_weight=0.05 and ctc_weight=0.5? Note that ESPnet2 doesn't implement unigram lsm.
> 
> Could you advice me @sw005320 ?
> I'm not sure whether this is a sign of some bugs. I don't know the effect of lsm and what kind of data babel is.

I've tested with label smoothing with the default initialization (which is not good) but yet test it on xavier_uniform. I can try it soon (it should not be a long time)

**Comment by kamo-naoyuki** [comment]
Thank you @ftshijt much. 
Perhaps, the rnn implementation, not vgg_rnn, has some problems...? I didn't test rnn version enough. 
I will also test in another task tomorrow and review the source code .

**Comment by ftshijt** [comment]
> Thank you @ftshijt much.
> Perhaps, the rnn implementation, not vgg_rnn, has some problems...? I didn't test rnn version enough.
> I will also test in another task tomorrow and review the source code .

I just got my result with label smoothing and observed 5% percent improvement (CER&WER). Besides, the acc line also seems better. An interesting trend is that the WER is always lower than the espnet1 version, while CER is not (ami ad commonvoice also have this kind of results, though not as severe) 

## asr_lsm_torch14_xavier_init
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_devdecode_asr_model_valid.loss.best|6623|56489|30.7|53.8|15.5|4.5|73.8|88.8|
|decode_eval_203decode_asr_model_valid.loss.best|11354|91411|29.7|53.8|16.5|4.7|75.0|89.4|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_devdecode_asr_model_valid.loss.best|6623|245555|59.2|14.8|26.0|4.0|44.8|89.4|
|decode_eval_203decode_asr_model_valid.loss.best|11354|397014|57.5|15.4|27.1|4.0|46.4|90.0|

acc

**Comment by kamo-naoyuki** [comment]
The number of words, 91411, doesn't equal to the number in https://github.com/espnet/espnet/blob/master/egs/babel/asr1/RESULTS . Characters number also differs.

Could you investigate why it's not same? Nlsyms?

**Comment by ftshijt** [comment]
> The number of words, 91411, doesn't equal to the number in https://github.com/espnet/espnet/blob/master/egs/babel/asr1/RESULTS . Characters number also differs.
> 
> Could you investigate why it's not same? Nlsyms?

I just checked it. The problem is due to "<unk>". In espnet2, it seems that if we state to use a nlsym.txt without <unk> in it, the reference text will include the "<unk>".

**Comment by kamo-naoyuki** [comment]
Babel includes `<unk>` in the reference text originally? OR, is it converted at some where in the recipe?
Exactly `<unk>`? Isn't is `<UNK>`?

I couldn't find where `<unk>` is removed in espnet1, but to remove it, we should just add it in nlsyms.text 
 @sw005320 Could you tell me how we should handle `<unk>`?

**Comment by ftshijt** [comment]
> Babel includes `<unk>` in the reference text originally? OR, is it converted at some where in the recipe?
> Exactly `<unk>`? Isn't is `<UNK>`?
> 
> I couldn't find where `<unk>` is removed in espnet1, but to remove it, we should just add it in nlsyms.text
> @sw005320 Could you tell me how we should handle `<unk>`?

Yes, the Babel includes `<unk>` originally and it is exactly `<unk>`. Since we manually add the `<unk>` to the token_list, I was figuring do not add the `<unk>` again. In espnet1, the `<unk>` is processed like other nlsyms such as `<noise>`, `<hes>`.

For current experiments, the percentage of `<unk>` is not large. I'm not sure if it will affect a large extent of the result. But if it is necessary to remove the `<unk>`, I can redo the experiments.

**Comment by kamo-naoyuki** [comment]
Maybe..., It's almost enough to recalculate error rate with new nlsyms. Could show me?

**Comment by ftshijt** [comment]
> Maybe..., It's almost enough to recalculate error rate with new nlsyms. Could show me?

As you wish, it's getting slightly better (and also the word/character is the same as espnet1)

## asr_lsm_torch14_xavier_init
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_devdecode_asr_model_valid.loss.best|6623|55847|31.1|53.9|15.0|5.0|74.0|88.6|
|decode_eval_203decode_asr_model_valid.loss.best|11354|90303|30.0|53.9|16.1|5.3|75.3|89.3|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_devdecode_asr_model_valid.loss.best|6623|242345|59.9|14.7|25.4|4.4|44.6|89.4|
|decode_eval_203decode_asr_model_valid.loss.best|11354|391474|58.3|15.2|26.6|4.5|46.2|90.0|

**Comment by kamo-naoyuki** [comment]
Good. I'm not sure, but maybe it's acceptable for now and not bug.

File: egs2/babel/asr1/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+from this directory ...
+
+`./setup_experiment.sh <expname>`
+
+This will copy all the necessary files for running an experiment to the created
+directory ../expname
+
+Possibly the most important of these files is 
+
+`conf/lang.conf`
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/README.md]: I didn't realized this file, but did this recipe write the database path to lang.conf in babel?
Hmm, kaldi uses really not consistent style between tasks...  Unbelievable...

Sorry, we must unify the style as db.sh, so could you change? Like BABEL_CANTONESE, ...

File: egs2/babel/asr1/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
+# BABEL RECIPE
+
+## Setting Up Experiments
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/README.md]: We should remove or modify this file.
Maybe we can remove them except this part:

```
### Examples
`./run.sh --langs "102" --recog "102"`
  This will create a network trained on 102 (assamese) and will test it on 102.

`./run.sh --langs "102 103" --recog " 102 103"`
  This will create a network trained on 102 and 103 (bengali) and tests on both.
```

File: egs2/babel/asr1/RESULTS.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<!-- Generated by scripts/utils/show_asr_result.sh -->
+# RESULTS
+## Environments
+- date: `Sun Mar  1 21:23:27 EST 2020`
+- python version: `3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0]`
+- espnet version: `espnet 0.6.0`
+- pytorch version: `pytorch 1.4.0`
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/RESULTS.md]: Could you merge this file into README.md? I'd like to show the result more easy.

File: egs2/babel/asr1/README.md (hunk 5 above)
Comment (ftshijt) [reply egs2/babel/asr1/README.md]: Screen Shot 2020-03-01 at 22 14 58
Yes, basically it has the dataset path, but in a very rough manner like this...

Comment (ftshijt) [reply egs2/babel/asr1/README.md]: Babel also employ lexicon file for initializing, can I put their paths in `db.sh` as well

Comment (kamo-naoyuki) [reply egs2/babel/asr1/README.md]: The shell variable in db.sh is supposed to correspond to the root directory of corpus which downloaded from original source as it is, e.g LDC.
I haven't get babel corpus, so I don't know what level is the root. Isn't the lexicon included at the same data corpus?

I guess as following:

```
# Cantonese
train_data_dir_101=$LDC2016S02/release-current/conversational/training
train_data_list_101=./conf/lists/101-cantonese/train.LimitedLP.list
train_data_dir_101_FLP=$LDC2016S02//release-current/conversational/training
train_data_list_101_FLP=./conf/lists/101-cantonese/train.FullLP.list
dev10h_data_dir_101=$LDC2016S02/release-current/conversational/dev
dev10h_data_list_101=./conf/lists/101-cantonese/dev.list
lexicon_file_101=$LDC2016S02/release-current/conversational/reference_materials/lexicon.sub-train1.txt
lexiconFlags_101="--romanized --oov <unk>"
```

( I just found where unk is inserted, the last line is )

Comment (ftshijt) [reply egs2/babel/asr1/README.md]: Gotcha! I will finish this format issue soon

File: egs2/babel/asr1/setup_experiment.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+if [ $# -ne 1 ]; then
+  echo >&2 "Usage: ./setup_experiment.sh <expname>"
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/setup_experiment.sh]: We don't need this file?

Comment (ftshijt) [reply egs2/babel/asr1/setup_experiment.sh]: Sure (sorry, it's a mistaken heritage from Kaldi

File: egs2/ami/asr1/local/data.sh [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	! hash BeamformIt && log "Missing BeamformIt, run 'cd ../../../tools/kaldi/tools; extras/install_beamformit.sh; cd -;'" && exit 1
+	local/ami_beamform.sh --cmd "${train_cmd}" --nj 20 ${nmics} ${AMI} ${PROCESSED_AMI_DIR}
+    else
+	PROCESSED_AMI_DIR=${AMI}
+    fi
+    local/ami_${base_mic}_data_prep.sh ${PROCESSED_AMI_DIR} ${mic}
+    # data augmentation
+    utils/perturb_data_dir_speed.sh 0.9 data/${mic}/train_orig data/ihm_tmp1
+    utils/perturb_data_dir_speed.sh 1.0 data/${mic}/train_orig data/ihm_tmp2    
+    utils/perturb_data_dir_speed.sh 1.1 data/${mic}/train_orig data/ihm_tmp3
```
Comment (kamo-naoyuki) [inline egs2/ami/asr1/local/data.sh]: Change to use --speed_perturb_factors
https://github.com/espnet/espnet/blob/24b07c73e0c3b98177d376cb7e0e5c936ed78a53/egs2/aishell/asr1/run.sh#L19-L36

File: egs2/ami/asr1/local/data.sh [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    rm -r data/${mic}/train_orig
+    utils/combine_data.sh --extra-files utt2uniq data/ihm/train_orig data/ihm_tmp1 data/ihm_tmp2 data/ihm_tmp3
+    
+    local/ami_${base_mic}_scoring_data_prep.sh ${PROCESSED_AMI_DIR} ${mic} dev
+    local/ami_${base_mic}_scoring_data_prep.sh ${PROCESSED_AMI_DIR} ${mic} eval
+    for dset in train dev eval; do
+	# changed the original AMI data structure in the Kaldi recipe to the following
+	utils/data/modify_speaker_info.sh --seconds-per-spk-max 30 data/${mic}/${dset}_orig data/${mic}_${dset}
+    done
```
Comment (kamo-naoyuki) [inline egs2/ami/asr1/local/data.sh]: Remove `utils/data/modify_speaker_info.sh` because we can't use kaldi command by default in espnet2,but modify_speaker_info requires  `wav-to-duration`. We don't need to modify speaker.

File: egs2/TEMPLATE/asr1/db.sh [hunk 11]
```diff
@@ -53,9 +80,35 @@ else
     JSUT=downloads
     TIMIT=
     VOXFORGE=downloads
+    AMI=
+    COMMONVOICE=
```
Comment (kamo-naoyuki) [inline egs2/TEMPLATE/asr1/db.sh]:
```suggestion
    COMMONVOICE=downloads
```

File: egs2/babel/asr1/conf/lang.conf [hunk 12]
```diff
@@ -0,0 +1,292 @@
+# A giant configurations file for all the BABEL languages
+# as well as some training configurations for training HMM-GMM systems
+# for obtaining phoneme level alignments if you really want to do that
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/conf/lang.conf]: This file is used?

File: egs2/babel/asr1/local/make_corpus_subset.sh [hunk 13]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash 
+
+# Copyright 2012  Johns Hopkins University (Author: Yenda Trmal)
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/local/make_corpus_subset.sh]: Did you modify anything in this file?
If no modification, then use symbolic links to original recipe for `local` script as possible.
(I can't compare all, so just making sure)

File: egs2/babel/asr1/local/run_all.sh [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+for x in 101-cantonese 102-assamese 103-bengali 104-pashto 105-turkish 106-tagalog 107-vietnamese 201-haitian 202-swahili 203-lao 204-tamil 205-kurmanji 206-zulu 207-tokpisin 404-georgian; do
+    langid=`echo $x | cut -f 1 -d"-"`
```
Comment (kamo-naoyuki) [inline egs2/babel/asr1/local/run_all.sh]: We don't need this file too

File: egs2/ami/asr1/conf/lm.yaml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+lm_conf:
+    unit: 1000   # 2 for character LMs
```
Comment (kamo-naoyuki) [inline egs2/ami/asr1/conf/lm.yaml]: Please follow the naming rules of #1617 to validate confguration files in CI.

Comment (kamo-naoyuki) [reply egs2/ami/asr1/conf/lm.yaml]: train_lm*.yaml or train_asr*.yaml

File: egs2/ami/asr1/local/data.sh (hunk 10 above)
Comment (ftshijt) [reply egs2/ami/asr1/local/data.sh]: do you mean we cannot use any utils scripts from kaldi?

Comment (kamo-naoyuki) [reply egs2/ami/asr1/local/data.sh]: I mean we can't use binary by default and we still use utils script and we can also still use fbank_pitch just as an option.
modify_speaker_info.sh invokes wav-to-duration and it's c++ command, so I'd like to avoid it.

**Comment by kamo-naoyuki** [comment]
LGTM. Really thank you a lot @ftshijt
